### PR TITLE
GraphChanged event

### PR DESCRIPTION
**Pull request by weitzhandler**
Addresses #171

### CODE REVIEW

File: Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs [hunk 1]
```diff
@@ -719,6 +719,23 @@ public void EntityChanged_Event_Should_Fire_When_Items_Added_Modified_Deleted()
...
 
+        [Fact]
+        public void GraphChanged_Event_Should_Fire_When_Graph_Items_Added_Modified_Deleted()
+        {
+          //Arrange
+          var database = new MockNorthwind();
+          var product = database.Products[0];
+          var changeTracker = new ChangeTrackingCollection<Product>(product);
+          var raised = false;
+          changeTracker.GraphChanged += (s, e) => raised = true;
```
Comment (tonysneed) [inline Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs]: I would prefer to keep EntityChanged, which fires when *any* entity in a graph of tracked objects has changed.

Comment (weitzhandler) [inline Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs]: What do you mean? Should we keep only one event that fires when the graph is changed?

**Comment by weitzhandler** [comment]
Reading [your comment](https://github.com/TrackableEntities/trackable-entities/issues/171#issuecomment-315548991), if I'm to implement this in the `SetTracking` ex. method, since the event is in the CTC:
1. How do I refer to the CTC from the `SetTracking` method, is `visitationHelper.OfType<ITrackingCollection>().First()` reliable?
2. Should I create a method (maybe internal) `RaiseEntityChanged` so that the event can be raised from the ex. method (events can only be raised in current class)?

**Comment by tonysneed** [comment]
Concerning subscribing to `EntityChanged`: There is also a `SetTracking` method in the CTC, which works in coordination with `SetTracking` in TrackableExtensions. The one in CTC can reference `this.EntityChanged`, and the one in TrackableExtensions references both the ref and collection property change trackers.

I don't think you have to worry about raising `EntityChanged`, because it is already being raised whenever an entity property is modified and when the entity is added or removed.

**Comment by weitzhandler** [comment]
Yes but what are my options to acquire the children.
As I see it, I have two options:

1. Pass a reference to the CTC via an argument in the `SetTracking` ex. method, and pass on all descending CTC's `EntityChanged` event to parent (via a `public`/`internal` `RaiseEntityChanged` method on the CTC
2. Have the `SetTracking` ex. method return an `IEnumerable<ITrackingCollection>` of all processed entities for the `CTC.SetTracking` to subscribe to their `EntityChanged` event.

I hope you have better ideas.

**Comment by tonysneed** [comment]
OK it looks like we need to change the signature of `TrackableExtensions.SetTracking` to accept an additional parameter: `EventHandler entityChanged`.  Then when a CTC reference is encountered, add it to the `EntityChanged` event:

```csharp
refChangeTracker.EntityChanged += entityChanged;
colProp.EntityCollection.EntityChanged += entityChanged;
```

Then in `ChangeTrackingCollection.SetTracking`:

```csharp
item.SetTracking(value, visitationHelper, oneToManyOnly, EntityChanged);
```

I don't know for sure, but I think this will result in the EntityChanged event of the topmost CTC getting fired whenever an entity is changed anywhere in the object graph.

**Comment by tonysneed** [comment]
PS. You need to do `+=` when tracking is being turned on, and `-=` when it is being turned off.

**Comment by weitzhandler** [comment]
The problem with this way is that it will work only if the handlers are attached before setting the tracking. We have to raise the event, not fire the handlers (which might join the party in a later stage thus neglected).

**Comment by weitzhandler** [comment]
The problem with this way is that it will work only if the handlers are attached before setting the tracking. We have to raise the event, not fire the handlers (which might join the party in a later stage hence will be
 neglected).

This will not work:

    //handler is null when SetTracking is called
    var changeTracker = new ChangeTrackingCollection<Order>(order);

    //handler wasn't attached to graph entities
    changeTracker.EntityChanged += handler;

~~~Maybe we should expand the `EntityChanged` event, so when a handler is add/removed, it calls `SetTracking` all over again, or a different method that only attaches/unattaches the handlers without touching the tracking. Not sure if even possible. Anyway yikes.~~~

**Comment by tonysneed** [comment]
Can you create a commit with your attempt and then push it? Then I can pull and try it out. Thanks!

File: Source/TrackableEntities.Client/ChangeTrackingCollection.cs [hunk 2]
```diff
@@ -21,16 +21,31 @@ public class ChangeTrackingCollection<TEntity> : ObservableCollection<TEntity>,
...
         // Deleted entities cache
         private readonly Collection<TEntity> _deletedEntities = new Collection<TEntity>();
-
+        
+        //private event EventHandler _EntityChanged;
         /// <summary>
         /// Event for when an entity in the collection graph has changed its tracking state.
         /// </summary>
         public event EventHandler EntityChanged;
+    //{
```
Comment (weitzhandler) [inline Source/TrackableEntities.Client/ChangeTrackingCollection.cs]: Notice this.

**Comment by tonysneed** [comment]
I'll check it out ...

File: Source/TrackableEntities.Client/ChangeTrackingCollection.cs [hunk 3]
```diff
@@ -90,15 +105,15 @@ public bool Tracking
             }
             set
             {
-                SetTracking(value, new ObjectVisitationHelper(), false);
+                SetTracking(value, new ObjectVisitationHelper(), false, EntityChanged);
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/ChangeTrackingCollection.cs]: Don't pass `EntityChanged` here.

File: Source/TrackableEntities.Client/ChangeTrackingCollection.cs [hunk 4]
```diff
@@ -90,15 +105,15 @@ public bool Tracking
...
             }
         }
 
         /// <summary>
         /// For internal use.
         /// Turn change-tracking on and off with proper circular reference checking.
         /// </summary>
-        public void SetTracking(bool value, ObjectVisitationHelper visitationHelper, bool oneToManyOnly)
+        public void SetTracking(bool value, ObjectVisitationHelper visitationHelper, bool oneToManyOnly, EventHandler entityChanged)
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/ChangeTrackingCollection.cs]: Don't add `entityChanged` to this method.

File: Source/TrackableEntities.Client/ITrackingCollection.cs [hunk 5]
```diff
@@ -27,8 +27,8 @@ public interface ITrackingCollection : ICollection
         /// <summary>
         /// For internal use.
         /// </summary>
-        void SetTracking(bool value, Common.ObjectVisitationHelper visitationHelper,
-            bool oneToManyOnly);
+        void SetTracking(bool value, Common.ObjectVisitationHelper visitationHelper, 
+            bool oneToManyOnly, EventHandler entityChanged);
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/ITrackingCollection.cs]: Don't change the interface.

File: Source/TrackableEntities.Client/TrackableExtensions.cs [hunk 6]
```diff
@@ -37,7 +42,8 @@ internal static class TrackableExtensions
                         if (refChangeTracker != null)
                         {
                             // Set tracking on ref prop change tracker
-                            refChangeTracker.SetTracking(enableTracking, visitationHelper, oneToManyOnly);
+                            refChangeTracker.SetTracking(enableTracking, visitationHelper, oneToManyOnly, entityChanged);
+                            refChangeTracker.SetHandler(enableTracking, entityChanged);
                         }
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/TrackableExtensions.cs]: Just call `SetHandler` here.

File: Source/TrackableEntities.Client/TrackableExtensions.cs [hunk 7]
```diff
@@ -47,9 +53,22 @@ internal static class TrackableExtensions
                 {
                     bool isOneToMany = !IsManyToManyChildCollection(colProp.EntityCollection);
                     if (!oneToManyOnly || isOneToMany)
-                        colProp.EntityCollection.SetTracking(enableTracking, visitationHelper, oneToManyOnly);
+                    {
+                        colProp.EntityCollection.SetTracking(enableTracking, visitationHelper, oneToManyOnly, entityChanged);
+                        colProp.EntityCollection.SetHandler(enableTracking, entityChanged);
+                    }
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/TrackableExtensions.cs]: Don't pass `entityChanged` to `SetTracking.` Just call `SetHandler` here.

File: Source/TrackableEntities.Client/TrackableExtensions.cs [hunk 8]
```diff
@@ -60,7 +79,7 @@ internal static class TrackableExtensions
         /// <param name="visitationHelper">Circular reference checking helper</param>
         /// <param name="isManyToManyItem">True is an item is treated as part of an M-M collection</param>
         public static void SetState(this ITrackable item, TrackingState state, ObjectVisitationHelper visitationHelper,
-            bool isManyToManyItem = false)
+                bool isManyToManyItem = false)
         {
             ObjectVisitationHelper.EnsureCreated(ref visitationHelper);
```
Comment (tonysneed) [inline Source/TrackableEntities.Client/TrackableExtensions.cs]: The visitor should prevent recursion.  We can circle back to take a closer look at this.

**Comment by tonysneed** [comment]
OK, so I made some comments on the code.  Let me know if they make sense .. I haven't implemented them, so they may or may not be valid. :)  It should not matter when the handlers are attached.

**Comment by weitzhandler** [comment]
1. You didn't say anything about my [previous comment](https://github.com/TrackableEntities/trackable-entities/pull/173#issuecomment-315617759)
2. Not changing the pattern in the CTC (only in the ex. method) will only pass the handler on to the first layer of children, but not to the navigation properties (which is turned on by calling their own CTC's `SetTracking`. I'll improve the test to cover that.

**Comment by tonysneed** [comment]
> Not changing the pattern in the CTC (only in the ex. method) will only pass the handler on to the first layer of children, but not to the navigation properties

`Set Tracking` in `TrackableExtensions` walks down the graph getting all the downstream CTC's.  The idea is that the handlers are chained.  Maybe the best way is for me to implement the way I'm describing and share it with you in a PR.  I could be wrong, but let me try that .  It's easier for me to demonstrate than just theorizing. :)

**Comment by weitzhandler** [comment]
The `SetTracking` ex. method, calls the nav. properties' `SetTracking`, which is the `CTC`'s, not the ex. method, thus, the parent `CTC`'s `EntityChanged` is neglected.

**Comment by tonysneed** [comment]
> The SetTracking ex. method, calls the nav. properties' SetTracking of the CTC

That's not how it's supposed to work .. let me demonstrate that.  I'll do a PR that shows what I mean ...

**Comment by weitzhandler** [comment]
You can use the same [branch](https://github.com/weitzhandler/trackable-entities/tree/graph-changed-event)

**Comment by tonysneed** [comment]
OK, so you gave me push rights?

**Comment by weitzhandler** [comment]
Yes

**Comment by tonysneed** [comment]
Can you resend the invite?  I clicked decline by mistake. ;)

**Comment by weitzhandler** [comment]
Done

**Comment by tonysneed** [comment]
Working on this, but I have to put the kids to bed first. :)

**Comment by weitzhandler** [comment]
Haha, mine are sleeping tight, they're soon to get up (hope not too soon)...

File: Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs [hunk 9]
```diff
@@ -719,6 +719,28 @@ public void EntityChanged_Event_Should_Fire_When_Items_Added_Modified_Deleted()
...
+          var changeCount = 0;
+
+          var database = new MockNorthwind();
+          var order = database.Orders[0];
+          var changeTracker = new ChangeTrackingCollection<Order> { order };
+          changeTracker.EntityChanged += (s, e) => changeCount++;
+          changeTracker.Tracking = true;
+
+          // Act
+          order.OrderDetails.Add(new OrderDetail());
```
Comment (tonysneed) [inline Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs]: So add this code to test the event on the parent:

```csharp
order.OrderDate = order.OrderDate.AddDays(1);
```

File: Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs [hunk 10]
```diff
@@ -719,6 +719,28 @@ public void EntityChanged_Event_Should_Fire_When_Items_Added_Modified_Deleted()
...
+          order.Customer.CustomerName = "Tony sneed";
+          order.OrderDetails.RemoveAt(0);
+          order.OrderDetails[0].Product.Category.CategoryName = "Lorem ipsum";
```
Comment (tonysneed) [inline Source/Tests/TrackableEntities.Client.Tests/ChangeTrackingCollectionTests.cs]: Just need to figure out why change to `Category` on line 738 does not fire the EntityChanged event.  See if we're not subscribing to it.

File: Source/TrackableEntities.Client/TrackableExtensions.cs (hunk 8 above)
Comment (tonysneed) [inline Source/TrackableEntities.Client/TrackableExtensions.cs]: Ignore my previous comment here .. looks like the only change was in the spacing.  Can you please discard the spacing changes? :-)

**Comment by tonysneed** [comment]
It looks like I've run out of time on this for now, and I won't be able to work on it myself for a while.  So I think the next step here would be to add the [line](https://github.com/TrackableEntities/trackable-entities/pull/173#discussion_r127701071) to the test that covers the root entity.  That will determine if the root entity is included.

Then you can figure out why the [change](https://github.com/TrackableEntities/trackable-entities/pull/173#discussion_r127701493) to `Category.CategoryName` is not firing the event.

**Comment by weitzhandler** [comment]
I'd like to get this one too.
Very important to me having an even raised whenever anything changes in the entire graph.

**Comment by tonysneed** [comment]
Thanks for working on this .. I'll plan on getting it into the 2.5.6 release.

**Comment by tonysneed** [comment]
I'm getting a failing test on this: `EntityChanged_Event_Should_Fire_When_Graph_Items_Added_Modified_Deleted`
Xunit.Sdk.EqualException
Assert.Equal() Failure
Expected: 4
Actual:   3

Maybe `CategoryChangeTracker` is not tracking changes?

**Comment by weitzhandler** [comment]
I'm afraid it's not enough changing the signature in the extensions, we also need to change the signature of `SetTracking` in the CTC to accept an handler for the child references when called recursively.

That's why these two don't raise the event:

    order.OrderDetails[0].Product.Discontinued = !order.OrderDetails[0].Product.Discontinued;
    order.OrderDetails[0].Product.Category.CategoryName = "Lorem ipsum";

**Comment by tonysneed** [comment]
I'll have a look.

**Comment by tonysneed** [comment]
Can you try implementing your proposal, and we can see if it works?

**Comment by weitzhandler** [comment]
How would you prefer implementing it? Should I change the existing `SetTracking` method in the interface? Add a new one? I now you wouldn't like changing the interface, that's why I wanted to introduce it as a separate `GraphChanged` event.

**Comment by tonysneed** [comment]
I figured it out .. have a look at PR #186 Recursive EntityChanged event in ChangeTrackingCollection.

**Pull request closed without merging**